server/tso/allocator_manager.go

```diff
@@ -430,33 +516,108 @@ func (am *AllocatorManager) allocatorPatroller(serverCtx context.Context) {
+			am.mu.clusterDCLocations[dcLocation] = &dcLocationInfo{
+				serverIDs: serverIDs,
+				suffix:    -1,
+			}
+		}
+	}
+	// Only leader can write the TSO suffix to etcd in order to make it consistent in the cluster
+	if am.member.IsStillLeader() {
+		for dcLocation, info := range am.mu.clusterDCLocations {
+			if info.suffix > 0 {
+				continue
+			}
+			suffixNum, err := am.getOrCreateLocalTSOSuffix(dcLocation)
+			if err != nil {
+				log.Warn("get or create the local tso suffix failed", zap.String("dc-location", dcLocation), errs.ZapError(err))
+				continue
+			}
+			am.mu.clusterDCLocations[dcLocation].suffix = suffixNum
+		}
+	}
+}
+
+// getOrCreateLocalTSOSuffix will check whether we have the Local TSO suffix written into etcd.
+// If not, it will write a number into etcd according to the its joining order.
+// If yes, it will just return the previous persisted one.
+func (am *AllocatorManager) getOrCreateLocalTSOSuffix(dcLocation string) (int32, error) {
+	// Try to get the suffix from etcd
+	resp, err := etcdutil.EtcdKVGet(
+		am.member.Client(),
+		am.GetLocalTSOSuffixPathPrefix(),
+		clientv3.WithPrefix())
+	if err != nil {
+		return -1, err
+	}
+	var maxSuffixNum int64
 	for _, kv := range resp.Kvs {
-		// The key will contain the member ID and the value is its dcLocation
-		serverPath := strings.Split(string(kv.Key), "/")
-		dcLocation := string(kv.Value)
-		// Get serverID from serverPath, e.g, /pd/dc-location/1232143243253 -> 1232143243253
-		serverID, err := strconv.ParseUint(serverPath[len(serverPath)-1], 10, 64)
+		suffix, err := strconv.ParseInt(string(kv.Value), 10, 32)
 		if err != nil {
-			log.Warn("get server id and dcLocation from etcd failed, invalid server id",
-				zap.Any("splitted-serverPath", serverPath),
-				zap.String("dc-location", dcLocation),
-				errs.ZapError(err))
-			continue
+			return -1, err
+		}
+		splittedKey := strings.Split(string(kv.Key), "/")
+		curDCLocation := splittedKey[len(splittedKey)-1]
+		// If we already have the suffix persistted in etcd before,
+		// just use it as the result directly.
+		if curDCLocation == dcLocation {
+			return int32(suffix), nil
 		}
-		am.mu.clusterDCLocations[dcLocation] = append(am.mu.clusterDCLocations[dcLocation], serverID)
+		if suffix > maxSuffixNum {
+			maxSuffixNum = suffix
+		}
+	}
+	maxSuffixNum++
+	localTSOSuffixKey := am.GetLocalTSOSuffixPath(dcLocation)
+	// The Local TSO suffix is determined by the joining order of this dc-location.
+	localTSOSuffixValue := strconv.FormatInt(maxSuffixNum, 10)
+	txnResp, err := kv.NewSlowLogTxn(am.member.Client()).
+		If(clientv3.Compare(clientv3.CreateRevision(localTSOSuffixKey), "=", 0)).
+		Then(clientv3.OpPut(localTSOSuffixKey, localTSOSuffixValue)).
+		Commit()
+	if err != nil {
+		return -1, errs.ErrEtcdTxn.Wrap(err).GenWithStackByCause()
+	}
+	// If the etcd txn failed, it means there is already a suffix written previously.
+	if !txnResp.Succeeded {
+		log.Warn("write local tso suffix into etcd failed",
```

Comment:
Will print this log periodically？

Reply:
It will print the log periodically only if the previous writing fails.

Reply:
```
// If the etcd txn failed, it means there is already a suffix written previously.
```
I think the comment says the previous write is successful.

Reply:
Fixed.